### PR TITLE
Use MapRootV1 in Key Transparency

### PR DESCRIPTION
**Pull request by gdbelvin**
This catches up with Trillian. Follow on to #949.

Now that `trillian.SignedMapRoot` does not directly expose fields, clients are (intentionally) forced to verify the signed map root before accessing it's fields.  This PR updates a few modules in KT to perform this verification where it was not happening before.

Changes:
-  Upgrade the `mutationclient` to verify `SignedMapRoot`s.  
   - Move the mutationclient[1]  into `client` which has the `mapVerifier`
      -  Autoconfiure the mutation client's poll period to be equal to the minimum delay between map revisions. 
- Helper methods for retrieving and verifying the latest signed map root.
- Update test vectors.

### CODE REVIEW

File: core/client/verify.go [hunk 1]
```diff
@@ -92,14 +92,15 @@ func (v *Verifier) Index(vrfProof []byte, domainID, appID, userID string) ([]byt
 //  - Verify signature.
 //  - Verify consistency proof from log.Root().
 //  - Verify inclusion proof.
+// Returns the new trusted log root.
```
Comment (pavelkalinnikov) [inline core/client/verify.go]: Not only log root?

File: core/client/verify.go [hunk 2]
```diff
@@ -110,53 +111,52 @@ func (v *Verifier) VerifyGetEntryResponse(ctx context.Context, domainID, appID,
...
-		return fmt.Errorf("VerifySignedMapRoot(): %v", err)
+		return nil, nil, fmt.Errorf("VerifySignedMapRoot(): %v", err)
 	}
+
 	Vlog.Printf("✓ Signed Map Head signature verified.")
+	if err := v.mapVerifier.VerifyMapLeafInclusion(mapRoot, leafProof); err != nil {
+		Vlog.Printf("✗ Sparse tree proof verification failed.")
+		return nil, nil, fmt.Errorf("VerifyMapLeafInclusion(): %v", err)
+	}
+	Vlog.Printf("✓ Sparse tree proof verified.")
```
Comment (pavelkalinnikov) [inline core/client/verify.go]: Nice idea with the checkmark/cross!

File: core/keyserver/epochs.go [hunk 3]
```diff
@@ -77,23 +77,21 @@ func (s *Server) GetEpoch(ctx context.Context, in *pb.GetEpochRequest) (*pb.Epoc
...
 	})
 	if err != nil {
-		glog.Errorf("GetEpoch(): GetSignedMapRootByRevision(%v, %v): %v", d.MapID, revision, err)
+		glog.Errorf("GetEpoch(): GetSignedMapRootByRevision(%v, %v): %v", d.MapID, mapRevision, err)
 		return nil, err
 	}
 
 	// MapRevisions start at 0. Log leaf indices starts at 0.
 	// MapRevision should be at least 1 since the Signer is
 	// supposed to create at least one revision on startup.
```
Comment (pavelkalinnikov) [inline core/keyserver/epochs.go]: nit: wrap 80?

File: core/monitor/monitor.go [hunk 4]
```diff
@@ -19,54 +19,60 @@ import (
...
 }
 
 // New creates a new instance of the monitor.
-func New(mClient pb.KeyTransparencyClient,
-	logVerifier *client.LogVerifier,
-	mapVerifier *client.MapVerifier,
+func New(cli *client.Client,
+	logVerifier *tclient.LogVerifier,
+	mapVerifier *tclient.MapVerifier,
 	signer *tcrypto.Signer,
```
Comment (pavelkalinnikov) [inline core/monitor/monitor.go]: nit: Make param order and struct field order consistent?

File: core/client/client.go [hunk 5]
```diff
@@ -94,52 +94,44 @@ func NewFromConfig(ktClient pb.KeyTransparencyClient, config *pb.Domain) (*Clien
...
+	minInterval, err := ptypes.Duration(config.MinInterval)
+	if err != nil {
+		return nil, err
+	}
+
+	return New(ktClient, config.DomainId, minInterval, ktVerifier), nil
 }
 
 // New creates a new client.
 // TODO(gbelvin): set retry delay.
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply core/client/client.go]: Removed

File: core/sequencer/sequencer.go [hunk 6]
```diff
@@ -81,6 +82,7 @@ type Sequencer struct {
 // New creates a new instance of the signer.
 func New(tlog tpb.TrillianLogClient,
 	logAdmin tpb.TrillianAdminClient,
+	mapAdmin tpb.TrillianAdminClient,
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply core/sequencer/sequencer.go]: Excellent suggestion

File: cmd/keytransparency-client/cmd/history.go [hunk 7]
```diff
@@ -58,14 +57,14 @@ and verify that the results are consistent.`,
...
+			slr, err := c.LatestSTH(ctx)
 			if err != nil {
 				return fmt.Errorf("GetEntry failed: %v", err)
 			}
 			if verbose {
-				fmt.Printf("Got current epoch: %v\n", smh.MapRevision)
+				fmt.Printf("Got current epoch: %v\n", slr.TreeSize-1)
 			}
-			end = smh.MapRevision
+			end = int64(slr.TreeSize) - 1
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply cmd/keytransparency-client/cmd/history.go]: Done

File: core/client/mutations.go [hunk 8]
```diff
@@ -12,12 +12,9 @@
...
 // limitations under the License.
 
-package mutationclient
+package client
 
-//
-// This file contains Monitor'c grpc client logic: poll mutations from the
-// kt-server mutations API and page if necessary.
-//
+// This file contains functions that download epochs and mutations
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply core/client/mutations.go]: Done

File: core/adminserver/admin_server.go [hunk 9]
```diff
@@ -269,22 +269,28 @@ func (s *Server) initialize(ctx context.Context, logTree, mapTree *tpb.Tree) err
...
+	if err != nil {
+		return fmt.Errorf("adminserver: NewMapVerifierFromTree(): %v", err)
+	}
+	mapRoot, err := mapVerifier.VerifySignedMapRoot(resp.GetMapRoot())
+	if err != nil {
+		return fmt.Errorf("adminserver: VerifySignedMapRoot(): %v", err)
+	}
 
 	// If the tree is empty and the map is empty,
 	// add the empty map root to the log.
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply core/adminserver/admin_server.go]: Golang does not have a specific line length that it encourages developers to use. 
Rather, it encourages developers to be sensible. In this case, I think these lines are clearer when not wrapped.

https://golang.org/doc/effective_go.html#formatting

File: core/client/client.go [hunk 10]
```diff
@@ -94,52 +94,44 @@ func NewFromConfig(ktClient pb.KeyTransparencyClient, config *pb.Domain) (*Clien
...
-	}
-	// Empty case.
-	if e.GetCommitted() == nil {
-		return nil, e.GetSmr(), nil
-	}
-
-	return e.GetCommitted().GetData(), e.GetSmr(), nil
+func (c *Client) GetEntry(ctx context.Context, userID, appID string, opts ...grpc.CallOption) ([]byte, *types.LogRootV1, error) {
+	e, slr, err := c.VerifiedGetEntry(ctx, appID, userID)
+	return e.GetCommitted().GetData(), slr, err
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply core/client/client.go]: The generated protobuf `Get*` functions work correctly even for `nil` receivers. It's quite nice!

File: core/client/gobindclient/client.go [hunk 11]
```diff
@@ -126,16 +126,11 @@ func GetEntry(ktURL, userID, appID string) ([]byte, error) {
...
 	if err != nil {
 		return nil, fmt.Errorf("GetEntry failed: %v", err)
 	}
-	// TODO(amarcedone): Consider returning or persisting smr it to verify consistency over time
-	_ = smr
-	//encodedSmr, err := proto.Marshal(smr)
-	//if err != nil {
-	//	return nil, fmt.Errorf("GetEntry failed: error serializing smr: %v", err)
-	//}
+	// TODO(amarcedone): Consider returning or persisting slr it to verify consistency over time
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply core/client/gobindclient/client.go]: Done

File: core/client/verify.go (hunk 1 above)
Comment (gdbelvin) [reply core/client/verify.go]: map root and log root.

File: core/client/verify.go (hunk 2 above)
Comment (gdbelvin) [reply core/client/verify.go]: UTF-8 source code!

File: core/keyserver/epochs.go (hunk 3 above)
Comment (gdbelvin) [reply core/keyserver/epochs.go]: It's a bit clearer unwrapped

File: core/monitor/monitor.go (hunk 4 above)
Comment (gdbelvin) [reply core/monitor/monitor.go]: done

File: core/client/client.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply core/client/client.go]: Cool, thanks for double checking.

File: core/client/verify.go (hunk 2 above)
Comment (pavelkalinnikov) [reply core/client/verify.go]: Yeah, I meant the cross above in line 140 :)

File: core/keyserver/epochs.go (hunk 3 above)
Comment (pavelkalinnikov) [reply core/keyserver/epochs.go]: Ack.

File: core/adminserver/admin_server.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply core/adminserver/admin_server.go]: Ack.

**Pull request merged**